File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        pair<int, int> max_element;
        map<int,int> my_map;
        map<int,int>::iterator itr;

        max_element.second = 0;

        for (size_t i = 0; i < nums.size(); i++)
        {
            itr = my_map.find(nums[i]);
            if(itr != my_map.end()){
                itr->second++;
            }
            else{
                my_map[nums[i]] = 1;
            }
            
            if(my_map[nums[i]] > max_element.second)
            {
                max_element.second = my_map[nums[i]];
                max_element.first = nums[i];
            }
        }

        return max_element.first;
        
    }
};

File: 0169-majority-element/0169-majority-element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        num_and_count = [nums[0], 1] # number, count

        for i in range(1, len(nums)):
            if nums[i] != num_and_count[0]:
                num_and_count[1] -= 1
                
                if num_and_count[1] == 0:
                    num_and_count = [nums[i], 1]
            else:
                num_and_count[1] += 1

        return num_and_count[0]

        